File: scenarios/csharp/dotnetcore/captioning/captioning/CaptionHelper.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.CognitiveServices.Speech;
using Microsoft.CognitiveServices.Speech.Intent;
using Microsoft.CognitiveServices.Speech.Translation;
using Microsoft.CognitiveServices.Speech.Transcription;
using System.Collections.Generic;
using System.Globalization;

namespace Captioning
{
    public struct Caption
    {
        public string? Language;
        public int Sequence;
        public TimeSpan Begin;
        public TimeSpan End;
        public string Text;
    }

    public class CaptionHelper
    {
        public static IEnumerable<Caption> GetCaptions(string? language, int maxWidth, int maxHeight, IEnumerable<object> results)
        {
            var helper = new CaptionHelper(language, maxWidth, maxHeight, results);
            return helper.GetCaptions();
        }

        public List<string> LinesFromText(string text)
        {
            var retval = new List<string>();
            
            var index = 0;
            while (index < text.Length)
            {
                index = SkipSkippable(text, index);

                int lineLength = GetBestWidth(text, index);
                retval.Add(text.Substring(index, lineLength).Trim());
                index = index + lineLength;
            }
            
            return retval;
        }

        public CaptionHelper(string? language, int maxWidth, int maxHeight, IEnumerable<object> results)
        {
            this._language = language;

            this._maxWidth = maxWidth;
            this._maxHeight = maxHeight;
            this._results = results;

            // consider adapting to use http://unicode.org/reports/tr29/#Sentence_Boundaries
            var iso639 = _language?.Split('-').FirstOrDefault();
            this._firstPassTerminators = iso639 switch
            {
                "zh" => new string[]{"，", "、", "；", "？", "！", "?", "!", ",", ";"},
                _ => new string[]{"?", "!", ",", ";"}
            };
            this._secondPassTerminators = iso639 switch
            {
                "zh" => new string[]{"。", " "},
                _ => new string[]{" ", "."}
            };

            if (maxWidth == UserConfig.defaultMaxLineLengthSBCS && iso639 == "zh")
            {
                this._maxWidth = UserConfig.defaultMaxLineLengthMBCS;
            }
        }

        public IEnumerable<Caption> GetCaptions()
        {
            EnsureCaptions();
            return _captions!;
        }

        private void EnsureCaptions()
        {
            if (_captions == null)
            {
                _captions = new List<Caption>();
                AddCaptionsForAllResults();
            }
        }

        private void AddCaptionsForAllResults()
        {
            foreach (RecognitionResult result in _results)
            {
                if (result.OffsetInTicks <= 0 || !IsFinalResult(result)) continue;

                var text = GetTextOrTranslation(result);
                if (string.IsNullOrEmpty(text)) continue;

                AddCaptionsForFinalResult(result, text);
            }
        }

        private string? GetTextOrTranslation(RecognitionResult result)
        {
            return result.Text;
            // 20220921 We do not use this for now because this sample
            // does not handle TranslationRecognitionResults.
            /*
            if (_language == null) return result.Text;

            return result is TranslationRecognitionResult translated && translated.Translations.Keys.Contains(_language)
                ? translated.Translations[_language]
                : null;
            */
        }

        private void AddCaptionsForFinalResult(RecognitionResult result, string text)
        {
            var captionStartsAt = 0;
            var captionLines = new List<string>();

            var index = 0;
            while (index < text.Length)
            {
                index = SkipSkippable(text, index);

                int lineLength = GetBestWidth(text, index);
                captionLines.Add(text.Substring(index, lineLength).Trim());
                index = index + lineLength;

                var isLastCaption = index >= text.Length;
                var maxCaptionLines = captionLines.Count >= _maxHeight;

                var addCaption = isLastCaption || maxCaptionLines;
                if (addCaption)
                {
                    var captionText = string.Join('\n', captionLines.ToArray());
                    captionLines.Clear();

                    var captionSequence = _captions!.Count + 1;
                    var isFirstCaption = captionStartsAt == 0;

                    (var captionBegin, var captionEnd) = isFirstCaption && isLastCaption
                        ? GetFullResultCaptionTiming(result)
                        : GetPartialResultCaptionTiming(result, text, captionText, captionStartsAt, index - captionStartsAt);

                    var caption = new Caption()
                    {
                        Language = _language,
                        Sequence = captionSequence,
                        Begin = captionBegin,
                        End = captionEnd,
                        Text = captionText
                    };
                    _captions.Add(caption);
                    
                    captionStartsAt = index;
                }
            }
        }

        private int GetBestWidth(string text, int startIndex)
        {
            var remaining = text.Length - startIndex;
            var bestWidth = remaining < _maxWidth
                ? remaining
                : FindBestWidth(_firstPassTerminators, text, startIndex);

            if (bestWidth < 0)
            {
                bestWidth = FindBestWidth(_secondPassTerminators, text, startIndex);
            }

            if (bestWidth < 0)
            {
                bestWidth = _maxWidth;
            }

            return bestWidth;
        }

        private int FindBestWidth(IEnumerable<string> terminators, string text, int startAt)
        {
            var remaining = text.Length - startAt;
            var checkChars = Math.Min(remaining, _maxWidth);

            var bestWidth = -1;
            foreach (var terminator in terminators)
            {
                var index = text.LastIndexOf(terminator, startAt + checkChars, checkChars);
                var width = index - startAt;
                if (width > bestWidth)
                {
                    bestWidth = width + terminator.Length;
                }
            }

            return bestWidth;
        }

        private int SkipSkippable(string text, int startIndex)
        {
            var index = startIndex;
            while (text.Length > index && text[index] == ' ')
                index++;

            return index;
        }

        private (TimeSpan captionBegin, TimeSpan captionEnd) GetFullResultCaptionTiming(RecognitionResult result)
        {
            var resultBegin = TimeSpan.FromTicks(result.OffsetInTicks);
            var resultEnd = resultBegin + result.Duration;
            return (resultBegin, resultEnd);
        }

        private (TimeSpan captionBegin, TimeSpan captionEnd) GetPartialResultCaptionTiming(RecognitionResult result, string text, string captionText, int captionStartsAt, int captionLength)
        {
            (var resultBegin, var resultEnd) = GetFullResultCaptionTiming(result);
            var resultDuration = resultEnd.Subtract(resultBegin);

            var textLength = text.Length;
            var partialBegin = resultBegin + resultDuration * captionStartsAt / textLength;
            var partialEnd = resultBegin + resultDuration * (captionStartsAt + captionLength) / textLength;

            return (partialBegin, partialEnd);
        }

        private static bool IsFinalResult(object result)
        {
            RecognitionResult? final = result as RecognitionResult;
            return final?.Reason == ResultReason.RecognizedSpeech ||
                   final?.Reason == ResultReason.RecognizedIntent ||
                   final?.Reason == ResultReason.TranslatedSpeech;
        }

        private readonly string? _language;
        private readonly string[] _firstPassTerminators;
        private readonly string[] _secondPassTerminators;

        private readonly int _maxWidth;
        private readonly int _maxHeight;
        private readonly IEnumerable<object> _results;

        private List<Caption>? _captions;
    }
}

File: scenarios/python/console/captioning/caption_helper.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
#

from datetime import date, datetime, time, timedelta
from typing import List, Optional, Tuple
import azure.cognitiveservices.speech as speechsdk # type: ignore
import helper

class Caption(object) :
    def __init__(self, language : Optional[str], sequence : int, begin : time, end : time, text : str) :
        self.language = language
        self.sequence = sequence
        self.begin = begin
        self.end = end
        self.text = text

def get_captions(language : Optional[str], max_width : int, max_height : int, results : List[dict]) -> List[Caption] :
    caption_helper = CaptionHelper(language, max_width, max_height, results)
    return caption_helper.get_captions()

class CaptionHelper(object) :
    def __init__(self, language : Optional[str], max_width : int, max_height : int, results : List[speechsdk.RecognitionResult]) :
        self._language = language
        self._max_width = max_width
        self._max_height = max_height
        self._results = results

        self._first_pass_terminators = ["?", "!", ",", ";"]
        self._second_pass_terminators = [" ", "."]

        self._captions : List[Caption] = []

        # consider adapting to use http://unicode.org/reports/tr29/#Sentence_Boundaries
        if self._language is not None :
            iso639 = self._language.split('-')[0]
            if "zh" == iso639.lower() :
                self._first_pass_terminators = ["，", "、", "；", "？", "！", "?", "!", ",", ";"]
                self._second_pass_terminators = ["。", " "]
                if (helper.DEFAULT_MAX_LINE_LENGTH_SBCS == self._max_width) :
                    self._max_width = helper.DEFAULT_MAX_LINE_LENGTH_MBCS

    def get_captions(self) -> List[Caption] :
        self.ensure_captions()
        return self._captions

    def ensure_captions(self) -> None :
        if not self._captions :
            self.add_captions_for_all_results()

    def add_captions_for_all_results(self) -> None :
        for result in self._results :
            if result.offset <= 0 or not self.is_final_result(result) :
                continue
            text = self.get_text_or_translation(result)
            if not text :
                continue
            self.add_captions_for_final_result(result, text)

    def get_text_or_translation(self, result : speechsdk.RecognitionResult) -> Optional[str] :
        return result.text

        # 20220921 We do not use this for now because this sample
        # does not handle TranslationRecognitionResults.
        #if not self._language :
        #    return result.text
        #if type(result) is speechsdk.TranslationRecognitionResult and self._language in result.Translations :
        #    return result.Translations[self._language]
        #else :
        #    return None

    def add_captions_for_final_result(self, result : speechsdk.RecognitionResult, text : str) -> None :
        caption_starts_at = 0
        caption_lines : List[str] = []
        index = 0
        while (index < len(text)) :
            index = self.skip_skippable(text, index)

            line_length = self.get_best_width(text, index)
            caption_lines.append(text[index:index + line_length].strip())
            index += line_length

            is_last_caption = index >= len(text)
            max_caption_lines = len(caption_lines) >= self._max_height

            add_caption = is_last_caption or max_caption_lines

            if add_caption :
                caption_text = '\n'.join(caption_lines)
                caption_lines.clear()

                caption_sequence = len(self._captions) + 1
                is_first_caption = 0 == caption_starts_at

                caption_begin_and_end : Tuple[time, time]
                if is_first_caption and is_last_caption :
                    caption_begin_and_end = self.get_full_caption_result_timing(result)
                else :
                    caption_begin_and_end = self.get_partial_result_caption_timing(result, text, caption_text, caption_starts_at, index - caption_starts_at)

                self._captions.append(Caption(self._language, caption_sequence, caption_begin_and_end[0], caption_begin_and_end[1], caption_text))
                
                caption_starts_at = index

    def get_best_width(self, text : str, start_index : int) -> int :
        remaining = len(text) - start_index
        best_width = remaining if remaining < self._max_width else self.find_best_width(self._first_pass_terminators, text, start_index)
        if (best_width < 0) :
            best_width = self.find_best_width(self._second_pass_terminators, text, start_index)
        if best_width < 0 :
            best_width = self._max_width
        return best_width

    def find_best_width(self, terminators : List[str], text : str, start_at : int) -> int :
        remaining = len(text) - start_at
        check_chars = min(remaining, self._max_width)
        best_width = -1
        for terminator in terminators :
            index = text.rfind(terminator, start_at, start_at + check_chars)
            width = index - start_at
            if width > best_width :
                best_width = width + len(terminator)
        return best_width

    def skip_skippable(self, text : str, start_index : int) -> int :
        index = start_index
        while len(text) > index and ' ' == text[index] :
            index += 1
        return index

    def get_full_caption_result_timing(self, result : speechsdk.RecognitionResult) -> Tuple[time, time] :
        begin = helper.time_from_ticks(result.offset)
        end = helper.time_from_ticks(result.offset + result.duration)
        return (begin, end)

    def get_partial_result_caption_timing(self, result : speechsdk.RecognitionResult, text : str, caption_text : str, caption_starts_at : int, caption_length : int) -> Tuple[time, time] :
        (result_begin, result_end) = self.get_full_caption_result_timing(result)
        result_duration = helper.subtract_times(result_end, result_begin)
        text_length = len(text)
        partial_begin = helper.add_time_and_timedelta(result_begin, result_duration * caption_starts_at / text_length)
        partial_end = helper.add_time_and_timedelta(result_begin, result_duration * (caption_starts_at + caption_length) / text_length)
        return (partial_begin, partial_end)

    def is_final_result(self, result : speechsdk.RecognitionResult) -> bool :
        return speechsdk.ResultReason.RecognizedSpeech == result.reason or speechsdk.ResultReason.RecognizedIntent == result.reason or speechsdk.ResultReason.TranslatedSpeech == result.reason

    def lines_from_text(self, text : str) -> List[str] :
        retval : List[str] = []
        index = 0
        while (index < len(text)) :
            index = self.skip_skippable(text, index)
            line_length = self.get_best_width(text, index)
            retval.append(text[index:index + line_length].strip())
            index += line_length
        return retval
